dart/common/Uri.cpp

```diff
@@ -196,12 +206,125 @@ void Uri::clear()
 }
 
 //==============================================================================
+bool Uri::fromString(const std::string& _input)
+{
+  // This is regex is from Appendix B of RFC 3986.
+  static regex uriRegex(
+    R"END(^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?)END",
+    regex::extended | regex::optimize);
+  static const size_t schemeIndex = 2;
+  static const size_t authorityIndex = 4;
+  static const size_t pathIndex = 5;
+  static const size_t queryIndex = 7;
+  static const size_t fragmentIndex = 9;
+
+  clear();
+
+  smatch matches;
+  if(!regex_match(_input, matches, uriRegex))
+    return false;
+
+  assert(matches.size() > schemeIndex);
+  const ssub_match& schemeMatch = matches[schemeIndex];
+  if(schemeMatch.matched)
+    mScheme = schemeMatch;
+
+  assert(matches.size() > authorityIndex);
+  const ssub_match& authorityMatch = matches[authorityIndex];
+  if(authorityMatch.matched)
+    mAuthority = authorityMatch;
+
+  assert(matches.size() > pathIndex);
+  const ssub_match& pathMatch = matches[pathIndex];
+  if(pathMatch.matched)
+  {
+    mPath = pathMatch;
+
+#ifdef _WIN32
+    if(mScheme && *mScheme == "file")
+    {
+      if(!mPath->empty() && mPath->at(0) == '/')
+        mPath = mPath->substr(1, mPath->size() - 1);
+    }
+#endif
+  }
```

Comment:
I would prefer to add a getter method that handles this conversion. Naming it `getPath` would be confusing, since it sounds like it would just return `mPath`. How about `getFilesystemPath`?

The key issue is that the word "path" could refer to two different concepts: (1) "the path component of a URI" or (2) "a path in the local filesystem". The `mPath` field on `Uri` specifically refers to definition (1), but definition (2) is more useful if you are trying to access a local file. These are interchangable on Linux, but differ by the leading `/` on Windows.